### PR TITLE
0206-upload Curve块存储IO链路零拷贝.md

### PR DESCRIPTION
**Pull request by xu-chaojie**

### CODE REVIEW

File: PrePaper/2023/0207-Curve块存储IO链路零拷贝.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+### 实现零拷贝
+
+要保证最终的日志落盘和数据落盘能够满足DMA write内存的对齐条件，需要我们在brpc收到网卡数据的第一时间，就将数据放到内存相应的对齐位置。只有这样，我们才能在后续不搬移（不拷贝）的情况下，使得最终需要落盘的数据满足内存对齐条件。因此，我们在修改基于rdma的brpc功能时，已经将brpc收到的数据attachment内存放到了满足DMA write内存对齐的位置。
+
+Follower段的数据转移路径基本上与Leader段是相似的，下面我们以Leader日志落盘场景为例，讲解CurveBS如何做到最终的日志和数据的对齐，从而实现零拷贝。
+以一个12K的IO为例:
+（1） 前面讲到我们在brpc收到的attachment时已经将其对齐到page，我们读取数据是一个iovec，且满足page对齐，如下图：
+
+![leader_log_write1.png](image/leader_log_write1.png)
```
Comment (wuhongsong) [inline PrePaper/2023/0207-Curve块存储IO链路零拷贝.md:70]: link is [invalid](javascript:;)

File: PrePaper/2023/0207-Curve块存储IO链路零拷贝.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+（2）接下来是日志编码过程，这一步会拼接上metasize + 序列化后的ChunkRequest，比如100字节。这一部分是还将作为attachment发送到follower端，这部分长度需要保证被4整除，这样Follower端收到的attachment才能是4字节对齐的。
+
+![leader_log_write2.png](image/leader_log_write2.png)
```
Comment (wuhongsong) [inline PrePaper/2023/0207-Curve块存储IO链路零拷贝.md:74]: ditto

File: PrePaper/2023/0207-Curve块存储IO链路零拷贝.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+（3）之后，在日志落盘前，还会在其头部添加raft log 的meta，假设为50字节。在这里，我们还需要避免产生read-merge-write， 我们还需要将这个IO按照块（4k或512）对齐。因此，我们在中间添加了一段填充（0）。结果如下：
+
+![leader_log_write3.png](image/leader_log_write3.png)
```
Comment (wuhongsong) [inline PrePaper/2023/0207-Curve块存储IO链路零拷贝.md:78]: ditto

File: PrePaper/2023/0207-Curve块存储IO链路零拷贝.md (hunk 1 above)
Comment (xu-chaojie) [reply PrePaper/2023/0207-Curve块存储IO链路零拷贝.md:70]: done

File: PrePaper/2023/0207-Curve块存储IO链路零拷贝.md (hunk 2 above)
Comment (xu-chaojie) [reply PrePaper/2023/0207-Curve块存储IO链路零拷贝.md:74]: done

File: PrePaper/2023/0207-Curve块存储IO链路零拷贝.md (hunk 3 above)
Comment (xu-chaojie) [reply PrePaper/2023/0207-Curve块存储IO链路零拷贝.md:78]: done

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**